sdks/python/apache_beam/runners/worker/statesampler_test.py

```diff
@@ -0,0 +1,98 @@
+#
+# Licensed to the Apache Software Foundation (ASF) under one or more
+# contributor license agreements.  See the NOTICE file distributed with
+# this work for additional information regarding copyright ownership.
+# The ASF licenses this file to You under the Apache License, Version 2.0
+# (the "License"); you may not use this file except in compliance with
+# the License.  You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+"""Tests for state sampler."""
+
+import logging
+import time
+import unittest
+
+from apache_beam.utils.counters import CounterFactory
+
+
+class StateSamplerTest(unittest.TestCase):
+
+  def test_basic_sampler(self):
+    # Set up state sampler.
+    counter_factory = CounterFactory()
+    sampler = statesampler.StateSampler('basic-', counter_factory,
+                                        sampling_period_ms=1)
+
+    # Run basic workload transitioning between 3 states.
+    sampler.start()
+    with sampler.scoped_state('statea'):
+      time.sleep(0.1)
+      with sampler.scoped_state('stateb'):
+        time.sleep(0.2 / 2)
+        with sampler.scoped_state('statec'):
+          time.sleep(0.3)
+        time.sleep(0.2 / 2)
+    sampler.stop()
+    sampler.commit_counters()
+
+    # Test that sampled state timings are within 5% of expected values.
```

Comment:
Yeah. It flakes about 2% of the time with 5% margin of error (and did so even without using locks). Full seconds would make it stabler, but make the test more expensive. Another option would be to require within 25%, and if not within 5% retry. 

Future PR (though I fixed the comment).